### PR TITLE
(draft/discussion) GRPO LoRA

### PR DESCRIPTION
**Pull request by ianbarber**
#### Context
What is the purpose of this PR? Is it to
- [x ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

#2421 - exploring a LoRA recipe. 

#### Changelog
What are the changes made in this PR?
* Add in new configs and recipe for lora distributed (though tested on a single device)
* Futz with the checkpointer

Putting this up early to get some input. Following the pattern from @RedTachyon I added the SFT config and the GRPO recipe. One immediate challenge was that checkpointing loading for adapters is only expected to be used when resuming: this case is more like a pipeline with SFT to lora first then a separate recipe with the same lora adapters. I ended up hacking around the checkpoint loading issues, but it doesn't feel particularly clean. 

On the positive side, by reducing the number of generations I got the recipe running in a single 24GB card, which is a good baseline for ease of access, not the fastest thing in the world, but it goes! 

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by ianbarber** [comment]
Updated this - I validated the recipe for single device, so figured it was just cleaner to do a single device diff. I reverted the checkpointing changes as the LoRA from SFT gets merged in and its cleaner just start it again. The recipe gets to a decent success rate during training - testing on 4 epochs running the gms8k task from lm evals gives:

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.5436|±  |0.0137|
|     |       |strict-match    |     5|exact_match|↑  |0.5421|±  |0.0137|
```
This is (clearly) eval-on-the-training data, but for reference the base llama 3B gets 0.2646 on flexible, and after one epoch .48. I was a little dubious whether LoRA would work here and it appears it does!

**Comment by pbontrager** [comment]
I just saw this PR has been hanging here since March! @ianbarber is this recipe not ready for review? Or did we just miss it?
